### PR TITLE
#85 - mongodb socobouser

### PR DESCRIPTION
**Pull request by flashback2k14**
@hetec i think it's ready for a PR. ;-P

### CODE REVIEW

**Comment by hetec** [comment]
If I try to register with
{
  "email": "test@test.de",
  "password": "password123",
  "role": "admin"
}

on localhost:8282/api/v1/auth/register

I always get a Validation Error -> Missing provider and so on

if I add the missing props

{
  "email": "test@test.de",
  "password": "password123",
  "role": "admin",
  "hasTermsAccepted": true,
  "provider": "Email",
  "imageUrl": "bla"
}

I get a 500 with a weird error message.

Can you help me with that, whats wrong with the request?

**Comment by hetec** [comment]
Error: error: Internal server error TypeError: Internal server error
    at Function.handleDbNotFound (/Users/patrickhebner/Documents/dev/projects/SocoboPlusPlus/server/dist/app/utils/error.utils.js:40:23)
    at SocoboUserRepository.<anonymous> (/Users/patrickhebner/Documents/dev/projects/SocoboPlusPlus/server/dist/socobouser/repositories/SocoboUser.repository.js:33:43)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/patrickhebner/Documents/dev/projects/SocoboPlusPlus/server/dist/socobouser/repositories/SocoboUser.repository.js:4:58)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)

**Comment by flashback2k14** [comment]
I look into it tomorrow. Haven't tested registration. ;-)

**Comment by hetec** [comment]
OK, thanks! Something similar happens if I try to request a user in my save recipe method. The User for the id (is in the db, I checked that) is alway null -> __transformResult throws an Error because there is no id in the result object (because is null)

**Comment by flashback2k14** [comment]
Could you please check in your code? Or id is now a string and not a number!?

**Comment by flashback2k14** [comment]
@hetec login and register should now work as expected. async/await is really nice! :-D

**Comment by hetec** [comment]
@flashback2k14 Yes works now thanks! And yes I like the async/await syntax, too!!! I will check the rest of the PR during the week

File: server/src/socobouser/schemas/SocoboUserSchema.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  },
+  provider: {
+    type: String
+  },
+  role: {
+    type: String
+  },
+  username: {
+    type: String
+  }
```
Comment (hetec) [inline server/src/socobouser/schemas/SocoboUserSchema.ts]: I use the option:
{
  timestamps: true
}

So I don't need to handle any created or last modified by my own. Mongoose will automatically handle these, and add the appropriate fields to the object.

File: server/src/app/utils/crypto.utils.ts [hunk 2]
```diff
@@ -4,42 +4,31 @@ import { ApiError, ComparePwResult, ERRORS } from "../index";
...
-          return reject(e);
-        }
-        resolve(new ComparePwResult(isMatch, user));
-      });
-    });
+  public comparePasswords = async (firstPw: string, secondPw: string): Promise<void> => {
+    try {
+      const result = await bcrypt.compare(firstPw, secondPw);
+      if (!result) {
+        throw new ApiError(ERRORS.AUTH_PW_MISSMATCH)
```
[the remark this answers is not in the document]
Comment (flashback2k14) [reply server/src/app/utils/crypto.utils.ts]: Yes, you are right. The ApiError(ERRORS.AUTH_PW_MISSMATCH) is added to the cause. It's only visible inside the logs. But I think we could delete the try catch and throw only the ApiError(ERRORS.AUTH_PW_MISSMATCH).

File: server/src/app/utils/error.utils.ts [hunk 3]
```diff
@@ -32,20 +32,26 @@ export class ErrorUtils {
...
-      return Promise.reject(e);
-    } else {
-      const e = new DbError(ERRORS.INTERNAL_SERVER_ERROR)
-        .addSource(source)
-        .addSourceMethod(method)
-        .addCause(error)
-        .addQuery(error.query);
-      return Promise.reject(e);
-    }
+    // if (ErrorUtils.notFound(error)) {
```
[the remark this answers is not in the document]
Comment (flashback2k14) [reply server/src/app/utils/error.utils.ts]: Nope. We could delete it.

File: server/src/socobouser/handlers/SocoboUser.handler.ts [hunk 4]
```diff
@@ -1,43 +1,50 @@
 import { Request, Response } from "express";
```
[the remark this answers is not in the document]
Comment (flashback2k14) [reply server/src/socobouser/handlers/SocoboUser.handler.ts]: Hmm. Yeah. I think we could change it to async await. :-)

File: server/src/socobouser/repositories/SocoboUser.repository.ts [hunk 5]
```diff
@@ -1,152 +1,160 @@
...
-        WHERE id=$1`;
-    return this._db.one(query, id, this._transformResult).catch((error: any) => {
-      return ErrorUtils.handleDbNotFound(
-        ERRORS.USER_NOT_FOUND, error, SocoboUserRepository.name, "getUserById(..)", "id", id.toString());
-      }
-    );
+  public getAll = async (): Promise<SocoboUser[] | DbError> => {
+    try {
+      const users = await this._socoboUserModel.find({});
+      return users.map(this._transformResult);
```
[the remark this answers is not in the document]
Comment (flashback2k14) [reply server/src/socobouser/repositories/SocoboUser.repository.ts]: We really need the _transformResult function only if we need functions from our model like getSigningInfo or clone. Maybe we need it later. Maybe not. :-)

File: server/src/socobouser/repositories/SocoboUser.repository.ts [hunk 6]
```diff
@@ -1,152 +1,160 @@
...
-  public getUserByEmail = (email: string): Promise<SocoboUser> => {
-    const query: string = "SELECT * FROM Socobo_User Where email=$1";
-    return this._db.one(query, email, this._transformResult).catch((error: any) => {
-      return ErrorUtils.handleDbNotFound (
-        ERRORS.USER_NOT_FOUND, error, SocoboUserRepository.name, "getUserByEmail(..)", "email", email);
-      }
-    );
+  public getUserById = async (id: Types.ObjectId): Promise<SocoboUser | DbError> => {
+    try {
+      const user = await this._socoboUserModel.findOne({_id: id});
```
[the remark this answers is not in the document]
Comment (flashback2k14) [reply server/src/socobouser/repositories/SocoboUser.repository.ts]: Ohh. Okay. I will change it. It reduce the imports count.

File: server/src/socobouser/schemas/SocoboUserSchema.ts (hunk 1 above)
Comment (flashback2k14) [reply server/src/socobouser/schemas/SocoboUserSchema.ts]: Yes, i know. But I liked my function. :-) we could change it if you like!?

File: server/src/socobouser/repositories/SocoboUser.repository.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (hetec) [reply server/src/socobouser/repositories/SocoboUser.repository.ts]: Ok. I think it's no problem to keep it.

File: server/src/socobouser/schemas/SocoboUserSchema.ts (hunk 1 above)
Comment (hetec) [reply server/src/socobouser/schemas/SocoboUserSchema.ts]: I just tried to update a user and missed to add the update type. This got me a 500. Maybe u should return a 400 Bad Request if the input json is invalid.

**Comment by hetec** [comment]
I can do an update like this:

{
	"fieldsToUpdate": {
	    "id": "59a30e9cba14db79b2382628",
	    "username": "",
	    "email": "john-doe",
	    "hasTermsAccepted": true,
	    "role": "USER",
	    "provider": "EMAIL",
	    "imageUrl": "socobo_data/placeholder/tmp_user_image.png"
	},
	"updateType": 0
}

This does not result in a validation error. But I think it should since username is required and email has a wrong format

**Comment by hetec** [comment]
I can also update single values in this way:

{
	"fieldsToUpdate": {
	    "username": "john-doe-test@test.test"
	},
	"updateType": 0
}

Is this ok?

If yes what do I need validation types for, if I'm also able to update single values in FULL mode?

**Comment by hetec** [comment]
I just tried to update a user and missed to add the update type. This got me a 500. Maybe u should return a 400 Bad Request if the input json is invalid. E.g. if the update type is to large

File: server/src/socobouser/schemas/SocoboUserSchema.ts (hunk 1 above)
Comment (hetec) [reply server/src/socobouser/schemas/SocoboUserSchema.ts]: No problem, this was just for information. I think it's not too bad if we use slightly different approaches, what do u think? Should we everywhere use exactly the same way of doing things?

Comment (flashback2k14) [reply server/src/socobouser/schemas/SocoboUserSchema.ts]: I think it's okay If it is not totally different.

**Comment by flashback2k14** [comment]
Hmm. I don't know why there is no validation errors. I must debug the update function and maybe bugfix it.

**Comment by flashback2k14** [comment]
@hetec i think the update function should work as expected now.

**Comment by hetec** [comment]
Hi, looks like it works for single properties like username and so. But if I try:

{
	"fieldsToUpdate": {
	    "username": "john-doe-test2@test.test",
	    "email": "john-doe@test.test",
	    "hasTermsAccepted": true,
	    "role": "ADMIN",
	    "provider": "EMAIL",
	    "imageUrl": "socobo_data/placeholder/tmp_user_image_test.png"
	},
	"updateType": 0
}

I get 

{
    "message": "There is no update type provided. This is required for updating a user.",
    "method": "checkUpdateType(..)",
    "source": "SocoboUserMiddleware"
}

What do I need to do to update all fields?

And sometime, for example if i use a string instead of a number for update type I get some 500 error. But this is maybe not so important

**Comment by flashback2k14** [comment]
there was a bug. Now it returns 
```json
{
    "message": "The provided update type is invalid!",
    "method": "checkUpdateBody(..)",
    "source": "SocoboUserMiddleware",
    "validationErrors": [
        "Password is missing"
    ]
}
```
and status code 400

the enums section in our readme tells you which fields needed for the different update types.

**Comment by hetec** [comment]
Looks good so far. I think there is still a little bug but I'm not sure. The readme tells me to provide 7 properties but your middleware restricts the property length to 6

switch (updateType) {
      case SocoboUserUpdateType.full:
        if (fieldLength <= 0 || **fieldLength > 6**) { errors.push("Invalid Property length"); }
        if (!fieldsToUpdate.hasOwnProperty("username")) { errors.push("Username is missing"); }
        if (!fieldsToUpdate.hasOwnProperty("email")) { errors.push("Email is missing"); }
        if (!fieldsToUpdate.hasOwnProperty("password")) { errors.push("Password is missing"); }
        if (!fieldsToUpdate.hasOwnProperty("imageUrl")) { errors.push("Image URL is missing"); }
        if (!fieldsToUpdate.hasOwnProperty("role")) { errors.push("Role is missing"); }
        if (!fieldsToUpdate.hasOwnProperty("provider")) { errors.push("Provider is missing"); }
        if (!fieldsToUpdate.hasOwnProperty("hasTermsAccepted")) { errors.push("Has Terms Accepted is missing"); }
        break;

So I always get a invalid property length error. Shouldn't this be 7?

I think after this is cleared we can merge the branch.

**Comment by flashback2k14** [comment]
Yeah, you are right. I haven't updated this after introducing the new terms update type. checkUpdateType also has this issue. I hope this is the last bug fix.

**Comment by flashback2k14** [comment]
i hope it's now ready for merge (after merging the other PR's into this PR).

**Comment by hetec** [comment]
I will check the food item PR soon!

**Comment by flashback2k14** [comment]
@hetec  Okay. Now all branches are merged into this branch. After this PR we could merge the other branches if we arrive the master branch (#84 #80 #75)

**Comment by flashback2k14** [comment]
@hetec could we please freeze the feature implementation and merge all open backend branches to bring the master branch up to date!?

**Pull request merged**